File: code/Java/src/main/java/Infrastructure/Timing/SmartTimer.java
package Infrastructure.Timing;

import Infrastructure.INotifier;

import java.util.ArrayList;
import java.util.Timer;
import java.util.TimerTask;

public class SmartTimer implements ISmartTimer {
    private int millisecondsToWait;
    private boolean isRunning = false;
    private Timer timer = new Timer();
    private TimerTask task;


    @Override
    public ISmartTimer waitMillisecond(int milliseconds) {
        reset();
        this.millisecondsToWait = milliseconds;
        return this;
    }

    @Override
    public ISmartTimer beforeDoing(INotifier notifier, ArrayList<String> datagrams) {

        task = new TimerTask() {
            @Override
            public void run() {
                notifier.notifyMessage(datagrams);
                reset();
            }
        };

        timer.schedule(task, millisecondsToWait, 1);
        return this;
    }

    private ISmartTimer reset() {
        if (task != null) {
            task.cancel();
        }
        return this;
    }

}

File: code/CSharp/ConnascentMarsRover/Source/Infrastructure/Timing/SmartTimer.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Threading;
using System.Threading.Tasks;

namespace Source.Infrastructure.Timing
{
    internal class SmartTimer : ISmartTimer
    {
        private int _millisecondsToWait;
        private Task _task;
        private CancellationToken _cancellationToken = CancellationToken.None;

        public ISmartTimer WaitMilliseconds(int milliseconds)
        {
            Reset();
            _millisecondsToWait = milliseconds;
            return this;
        }

        public ISmartTimer BeforeDoing(
            Action<List<string>> notifyMessage, List<string> packets)
        {
            _task = Task.Factory.StartNew(() =>
            {
                notifyMessage(packets);
                Reset();
            });

            _task.Wait(_millisecondsToWait, _cancellationToken);
            _cancellationToken = CancellationToken.None;
            return this;
        }

        private void Reset()
        {
            _cancellationToken = new CancellationToken();
        }
    }
}